File: go/pkg/validator/validator_test.go
package validator

import (
	"testing"

	"github.com/thirdkeyai/toolclad/pkg/manifest"
)

func intPtr(v int) *int         { return &v }
func floatPtr(v float64) *float64 { return &v }

func TestValidateArg(t *testing.T) {
	tests := []struct {
		name    string
		def     *manifest.ArgDef
		value   string
		want    string
		wantErr bool
	}{
		// --- string ---
		{
			name:  "string/valid",
			def:   &manifest.ArgDef{Name: "s", Type: "string"},
			value: "hello",
			want:  "hello",
		},
		{
			name:    "string/injection",
			def:     &manifest.ArgDef{Name: "s", Type: "string"},
			value:   "hello; rm -rf /",
			wantErr: true,
		},
		{
			name:  "string/pattern match",
			def:   &manifest.ArgDef{Name: "s", Type: "string", Pattern: `^[a-z]+$`},
			value: "hello",
			want:  "hello",
		},
		{
			name:    "string/pattern mismatch",
			def:     &manifest.ArgDef{Name: "s", Type: "string", Pattern: `^[a-z]+$`},
			value:   "HELLO",
			wantErr: true,
		},

		// --- integer ---
		{
			name:  "integer/valid",
			def:   &manifest.ArgDef{Name: "n", Type: "integer"},
			value: "42",
			want:  "42",
		},
		{
			name:    "integer/not a number",
			def:     &manifest.ArgDef{Name: "n", Type: "integer"},
			value:   "abc",
			wantErr: true,
		},
		{
			name:  "integer/min max valid",
			def:   &manifest.ArgDef{Name: "n", Type: "integer", Min: intPtr(1), Max: intPtr(100)},
			value: "50",
			want:  "50",
		},
		{
			name:    "integer/below min",
			def:     &manifest.ArgDef{Name: "n", Type: "integer", Min: intPtr(10), Max: intPtr(100)},
			value:   "5",
			wantErr: true,
		},
		{
			name:    "integer/above max",
			def:     &manifest.ArgDef{Name: "n", Type: "integer", Min: intPtr(1), Max: intPtr(10)},
			value:   "20",
			wantErr: true,
		},
		{
			name:  "integer/clamp low",
			def:   &manifest.ArgDef{Name: "n", Type: "integer", Min: intPtr(5), Max: intPtr(50), Clamp: true},
			value: "1",
			want:  "5",
		},
		{
			name:  "integer/clamp high",
			def:   &manifest.ArgDef{Name: "n", Type: "integer", Min: intPtr(5), Max: intPtr(50), Clamp: true},
			value: "100",
			want:  "50",
		},

		// --- port ---
		{
			name:  "port/valid",
			def:   &manifest.ArgDef{Name: "p", Type: "port"},
			value: "443",
			want:  "443",
		},
		{
			name:    "port/zero",
			def:     &manifest.ArgDef{Name: "p", Type: "port"},
			value:   "0",
			wantErr: true,
		},
		{
			name:    "port/negative",
			def:     &manifest.ArgDef{Name: "p", Type: "port"},
			value:   "-1",
			wantErr: true,
		},
		{
			name:    "port/too high",
			def:     &manifest.ArgDef{Name: "p", Type: "port"},
			value:   "70000",
			wantErr: true,
		},
		{
			name:    "port/not a number",
			def:     &manifest.ArgDef{Name: "p", Type: "port"},
			value:   "http",
			wantErr: true,
		},

		// --- boolean ---
		{
			name:  "boolean/true",
			def:   &manifest.ArgDef{Name: "b", Type: "boolean"},
			value: "true",
			want:  "true",
		},
		{
			name:  "boolean/false",
			def:   &manifest.ArgDef{Name: "b", Type: "boolean"},
			value: "false",
			want:  "false",
		},
		{
			name:  "boolean/True uppercase",
			def:   &manifest.ArgDef{Name: "b", Type: "boolean"},
			value: "True",
			want:  "true",
		},
		{
			name:    "boolean/invalid",
			def:     &manifest.ArgDef{Name: "b", Type: "boolean"},
			value:   "yes",
			wantErr: true,
		},

		// --- enum ---
		{
			name:  "enum/valid",
			def:   &manifest.ArgDef{Name: "e", Type: "enum", Allowed: []string{"ping", "syn", "service"}},
			value: "syn",
			want:  "syn",
		},
		{
			name:    "enum/invalid value",
			def:     &manifest.ArgDef{Name: "e", Type: "enum", Allowed: []string{"ping", "syn"}},
			value:   "udp",
			wantErr: true,
		},
		{
			name:    "enum/no allowed list",
			def:     &manifest.ArgDef{Name: "e", Type: "enum"},
			value:   "any",
			wantErr: true,
		},

		// --- scope_target ---
		{
			name:  "scope_target/ip",
			def:   &manifest.ArgDef{Name: "t", Type: "scope_target"},
			value: "192.168.1.1",
			want:  "192.168.1.1",
		},
		{
			name:  "scope_target/cidr",
			def:   &manifest.ArgDef{Name: "t", Type: "scope_target"},
			value: "10.0.0.0/24",
			want:  "10.0.0.0/24",
		},
		{
			name:  "scope_target/hostname",
			def:   &manifest.ArgDef{Name: "t", Type: "scope_target"},
			value: "example.com",
			want:  "example.com",
		},
		{
			name:    "scope_target/wildcard",
			def:     &manifest.ArgDef{Name: "t", Type: "scope_target"},
			value:   "*.example.com",
			wantErr: true,
		},
		{
			name:    "scope_target/injection",
			def:     &manifest.ArgDef{Name: "t", Type: "scope_target"},
			value:   "example.com; cat /etc/passwd",
			wantErr: true,
		},
		{
			name:    "scope_target/invalid",
			def:     &manifest.ArgDef{Name: "t", Type: "scope_target"},
			value:   "not a valid target!!!",
			wantErr: true,
		},

		// --- url ---
		{
			name:  "url/valid http",
			def:   &manifest.ArgDef{Name: "u", Type: "url"},
			value: "https://example.com/path",
			want:  "https://example.com/path",
		},
		{
			name:    "url/no scheme",
			def:     &manifest.ArgDef{Name: "u", Type: "url"},
			value:   "example.com",
			wantErr: true,
		},
		{
			name:  "url/scheme restriction pass",
			def:   &manifest.ArgDef{Name: "u", Type: "url", Schemes: []string{"https"}},
			value: "https://example.com",
			want:  "https://example.com",
		},
		{
			name:    "url/scheme restriction fail",
			def:     &manifest.ArgDef{Name: "u", Type: "url", Schemes: []string{"https"}},
			value:   "http://example.com",
			wantErr: true,
		},

		// --- path ---
		{
			name:  "path/valid_relative",
			def:   &manifest.ArgDef{Name: "f", Type: "path"},
			value: "wordlists/rockyou.txt",
			want:  "wordlists/rockyou.txt",
		},
		{
			name:    "path/absolute_rejected",
			def:     &manifest.ArgDef{Name: "f", Type: "path"},
			value:   "/usr/share/wordlists/rockyou.txt",
			wantErr: true,
		},
		{
			name:    "path/traversal",
			def:     &manifest.ArgDef{Name: "f", Type: "path"},
			value:   "/etc/../etc/passwd",
			wantErr: true,
		},

		// --- ip_address ---
		{
			name:  "ip_address/ipv4",
			def:   &manifest.ArgDef{Name: "ip", Type: "ip_address"},
			value: "192.168.1.1",
			want:  "192.168.1.1",
		},
		{
			name:  "ip_address/ipv6",
			def:   &manifest.ArgDef{Name: "ip", Type: "ip_address"},
			value: "::1",
			want:  "::1",
		},
		{
			name:    "ip_address/invalid",
			def:     &manifest.ArgDef{Name: "ip", Type: "ip_address"},
			value:   "not-an-ip",
			wantErr: true,
		},

		// --- cidr ---
		{
			name:  "cidr/valid",
			def:   &manifest.ArgDef{Name: "c", Type: "cidr"},
			value: "10.0.0.0/24",
			want:  "10.0.0.0/24",
		},
		{
			name:    "cidr/no slash",
			def:     &manifest.ArgDef{Name: "c", Type: "cidr"},
			value:   "10.0.0.0",
			wantErr: true,
		},
		{
			name:    "cidr/invalid",
			def:     &manifest.ArgDef{Name: "c", Type: "cidr"},
			value:   "999.999.999.0/24",
			wantErr: true,
		},

		// --- msf_options ---
		{
			name:  "msf_options/valid single",
			def:   &manifest.ArgDef{Name: "opts", Type: "msf_options"},
			value: "RHOSTS 192.168.1.1",
			want:  "RHOSTS 192.168.1.1",
		},
		{
			name:  "msf_options/valid multiple",
			def:   &manifest.ArgDef{Name: "opts", Type: "msf_options"},
			value: "RHOSTS 192.168.1.1; RPORT 445",
			want:  "RHOSTS 192.168.1.1; RPORT 445",
		},
		{
			name:    "msf_options/invalid key lowercase",
			def:     &manifest.ArgDef{Name: "opts", Type: "msf_options"},
			value:   "rhosts 192.168.1.1",
			wantErr: true,
		},
		{
			name:    "msf_options/missing value",
			def:     &manifest.ArgDef{Name: "opts", Type: "msf_options"},
			value:   "RHOSTS",
			wantErr: true,
		},
		{
			name:    "msf_options/pipe in value",
			def:     &manifest.ArgDef{Name: "opts", Type: "msf_options"},
			value:   "RHOSTS foo|bar",
			wantErr: true,
		},

		// --- credential_file ---
		{
			name:    "credential_file/absolute path",
			def:     &manifest.ArgDef{Name: "cred", Type: "credential_file"},
			value:   "/etc/shadow",
			wantErr: true,
		},
		{
			name:    "credential_file/path traversal",
			def:     &manifest.ArgDef{Name: "cred", Type: "credential_file"},
			value:   "creds/../../../etc/shadow",
			wantErr: true,
		},
		{
			name:    "credential_file/nonexistent",
			def:     &manifest.ArgDef{Name: "cred", Type: "credential_file"},
			value:   "nonexistent_file.txt",
			wantErr: true,
		},

		// --- duration ---
		{
			name:  "duration/plain seconds",
			def:   &manifest.ArgDef{Name: "d", Type: "duration"},
			value: "30",
			want:  "30",
		},
		{
			name:  "duration/minutes",
			def:   &manifest.ArgDef{Name: "d", Type: "duration"},
			value: "5m",
			want:  "5m",
		},
		{
			name:  "duration/hours and minutes",
			def:   &manifest.ArgDef{Name: "d", Type: "duration"},
			value: "1h30m",
			want:  "1h30m",
		},
		{
			name:  "duration/milliseconds",
			def:   &manifest.ArgDef{Name: "d", Type: "duration"},
			value: "500ms",
			want:  "500ms",
		},
		{
			name:    "duration/invalid",
			def:     &manifest.ArgDef{Name: "d", Type: "duration"},
			value:   "five minutes",
			wantErr: true,
		},
		{
			name:    "duration/empty",
			def:     &manifest.ArgDef{Name: "d", Type: "duration"},
			value:   "",
			wantErr: true,
		},

		// --- regex_match ---
		{
			name:  "regex_match/valid",
			def:   &manifest.ArgDef{Name: "r", Type: "regex_match", Pattern: `^[a-z0-9_]+$`},
			value: "hello_world",
			want:  "hello_world",
		},
		{
			name:    "regex_match/no match",
			def:     &manifest.ArgDef{Name: "r", Type: "regex_match", Pattern: `^[a-z0-9_]+$`},
			value:   "HELLO",
			wantErr: true,
		},
		{
			name:    "regex_match/no pattern",
			def:     &manifest.ArgDef{Name: "r", Type: "regex_match"},
			value:   "anything",
			wantErr: true,
		},

		// --- unknown type ---
		{
			name:    "unknown type",
			def:     &manifest.ArgDef{Name: "x", Type: "foobar"},
			value:   "anything",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidateArg(tt.def, tt.value)
			if tt.wantErr {
				if err == nil {
					t.Errorf("expected error, got nil (value=%q)", got)
				}
				return
			}
			if err != nil {
				t.Errorf("unexpected error: %v", err)
				return
			}
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestValidateArgWithCustomTypes(t *testing.T) {
	customTypes := map[string]*manifest.CustomTypeDef{
		"severity": {
			Base:    "enum",
			Allowed: []string{"low", "medium", "high", "critical"},
		},
		"bounded_int": {
			Base: "integer",
			Min:  intPtr(1),
			Max:  intPtr(100),
		},
	}

	tests := []struct {
		name    string
		def     *manifest.ArgDef
		value   string
		want    string
		wantErr bool
	}{
		{
			name:  "custom enum type valid",
			def:   &manifest.ArgDef{Name: "sev", Type: "severity"},
			value: "high",
			want:  "high",
		},
		{
			name:    "custom enum type invalid",
			def:     &manifest.ArgDef{Name: "sev", Type: "severity"},
			value:   "unknown",
			wantErr: true,
		},
		{
			name:  "custom integer type valid",
			def:   &manifest.ArgDef{Name: "num", Type: "bounded_int"},
			value: "50",
			want:  "50",
		},
		{
			name:    "custom integer type out of range",
			def:     &manifest.ArgDef{Name: "num", Type: "bounded_int"},
			value:   "200",
			wantErr: true,
		},
		{
			name:  "fallback to builtin type",
			def:   &manifest.ArgDef{Name: "p", Type: "port"},
			value: "443",
			want:  "443",
		},
		{
			name:    "custom type with invalid base",
			def:     &manifest.ArgDef{Name: "x", Type: "bad_custom"},
			value:   "anything",
			wantErr: true,
		},
	}

	// Add a custom type with invalid base for testing
	customTypes["bad_custom"] = &manifest.CustomTypeDef{Base: "nonexistent"}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidateArgWithCustomTypes(tt.def, tt.value, customTypes)
			if tt.wantErr {
				if err == nil {
					t.Errorf("expected error, got nil (value=%q)", got)
				}
				return
			}
			if err != nil {
				t.Errorf("unexpected error: %v", err)
				return
			}
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestNumberType(t *testing.T) {
	tests := []struct {
		name    string
		def     *manifest.ArgDef
		value   string
		wantErr bool
	}{
		{"valid float", &manifest.ArgDef{Name: "n", Type: "number"}, "0.5", false},
		{"negative", &manifest.ArgDef{Name: "n", Type: "number"}, "-3.14", false},
		{"NaN rejected", &manifest.ArgDef{Name: "n", Type: "number"}, "NaN", true},
		{"Inf rejected", &manifest.ArgDef{Name: "n", Type: "number"}, "Inf", true},
		{"non-numeric", &manifest.ArgDef{Name: "n", Type: "number"}, "abc", true},
		{
			"min_float ok",
			&manifest.ArgDef{Name: "n", Type: "number", MinFloat: floatPtr(0.0), MaxFloat: floatPtr(1.0)},
			"0.5",
			false,
		},
		{
			"min_float below",
			&manifest.ArgDef{Name: "n", Type: "number", MinFloat: floatPtr(0.0)},
			"-0.1",
			true,
		},
		{
			"max_float above",
			&manifest.ArgDef{Name: "n", Type: "number", MaxFloat: floatPtr(1.0)},
			"1.5",
			true,
		},
		{
			"falls back to int min/max",
			&manifest.ArgDef{Name: "n", Type: "number", Min: intPtr(1), Max: intPtr(10)},
			"5.5",
			false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := ValidateArg(tt.def, tt.value)
			if tt.wantErr && err == nil {
				t.Errorf("expected error, got nil")
			}
			if !tt.wantErr && err != nil {
				t.Errorf("unexpected error: %v", err)
			}
		})
	}
}

func TestScopeTargetHardening(t *testing.T) {
	tests := []struct {
		name    string
		value   string
		wantSub string // expected substring in error
	}{
		{"punycode rejected", "xn--example-9c.com", "punycode"},
		{"mixed-case punycode rejected", "XN--example-9c.com", "punycode"},
		{"non-ASCII rejected", "exаmple.com", "ASCII"}, // Cyrillic а
		{"specific traversal message", "../../etc/passwd", "traversal"},
		{"specific slash message", "etc/passwd", "'/'"},
	}
	def := &manifest.ArgDef{Name: "t", Type: "scope_target"}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := ValidateArg(def, tt.value)
			if err == nil {
				t.Fatalf("expected error for %q, got none", tt.value)
			}
			if !contains(err.Error(), tt.wantSub) {
				t.Errorf("error %q does not contain %q", err.Error(), tt.wantSub)
			}
		})
	}
}

func contains(haystack, needle string) bool {
	return len(haystack) >= len(needle) && (func() bool {
		for i := 0; i+len(needle) <= len(haystack); i++ {
			if haystack[i:i+len(needle)] == needle {
				return true
			}
		}
		return false
	})()
}

func TestCheckInjection(t *testing.T) {
	tests := []struct {
		value   string
		wantErr bool
	}{
		{"hello", false},
		{"hello world", false},
		{"192.168.1.1", false},
		{"hello;world", true},
		{"$(whoami)", true},
		{"foo|bar", true},
		{"foo&bar", true},
		{"foo`bar`", true},
		{"{}", true},
		{"a<b>c", true},
	}

	for _, tt := range tests {
		t.Run(tt.value, func(t *testing.T) {
			err := CheckInjection(tt.value)
			if tt.wantErr && err == nil {
				t.Error("expected error, got nil")
			}
			if !tt.wantErr && err != nil {
				t.Errorf("unexpected error: %v", err)
			}
		})
	}
}

File: js/test/validator.test.js
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { validateArg, checkInjection } from "../src/validator.js";

describe("checkInjection", () => {
  it("accepts clean strings", () => {
    checkInjection("hello-world");
    checkInjection("10.0.1.0/24");
    checkInjection("example.com");
  });

  it("rejects shell metacharacters", () => {
    assert.throws(() => checkInjection("foo;bar"), /Injection detected/);
    assert.throws(() => checkInjection("foo|bar"), /Injection detected/);
    assert.throws(() => checkInjection("$(whoami)"), /Injection detected/);
    assert.throws(() => checkInjection("foo`id`"), /Injection detected/);
    assert.throws(() => checkInjection("foo&bar"), /Injection detected/);
  });
});

describe("validateArg - string", () => {
  it("accepts a plain string", () => {
    const result = validateArg({ type: "string" }, "hello");
    assert.equal(result, "hello");
  });

  it("rejects injection in strings", () => {
    assert.throws(() => validateArg({ type: "string" }, "foo;rm -rf /"), /Injection/);
  });

  it("validates pattern constraint", () => {
    const def = { type: "string", pattern: "^[a-z]+$" };
    assert.equal(validateArg(def, "hello"), "hello");
    assert.throws(() => validateArg(def, "Hello123"), /does not match pattern/);
  });
});

describe("validateArg - integer", () => {
  it("accepts valid integers", () => {
    assert.equal(validateArg({ type: "integer" }, 42), 42);
    assert.equal(validateArg({ type: "integer" }, "10"), 10);
  });

  it("rejects non-integers", () => {
    assert.throws(() => validateArg({ type: "integer" }, "abc"), /Expected integer/);
    assert.throws(() => validateArg({ type: "integer" }, 3.5), /Expected integer/);
  });

  it("enforces min/max", () => {
    const def = { type: "integer", min: 1, max: 10 };
    assert.equal(validateArg(def, 5), 5);
    assert.throws(() => validateArg(def, 0), /below minimum/);
    assert.throws(() => validateArg(def, 11), /exceeds maximum/);
  });

  it("clamps when clamp=true", () => {
    const def = { type: "integer", min: 1, max: 10, clamp: true };
    assert.equal(validateArg(def, 0), 1);
    assert.equal(validateArg(def, 100), 10);
    assert.equal(validateArg(def, 5), 5);
  });
});

describe("validateArg - port", () => {
  it("accepts valid ports", () => {
    assert.equal(validateArg({ type: "port" }, 80), 80);
    assert.equal(validateArg({ type: "port" }, 65535), 65535);
    assert.equal(validateArg({ type: "port" }, "443"), 443);
  });

  it("rejects invalid ports", () => {
    assert.throws(() => validateArg({ type: "port" }, 0), /Invalid port/);
    assert.throws(() => validateArg({ type: "port" }, 70000), /Invalid port/);
    assert.throws(() => validateArg({ type: "port" }, "abc"), /Invalid port/);
  });
});

describe("validateArg - boolean", () => {
  it("accepts true/false strings", () => {
    assert.equal(validateArg({ type: "boolean" }, "true"), true);
    assert.equal(validateArg({ type: "boolean" }, "false"), false);
    assert.equal(validateArg({ type: "boolean" }, true), true);
  });

  it("rejects other values", () => {
    assert.throws(() => validateArg({ type: "boolean" }, "yes"), /Boolean must be/);
    assert.throws(() => validateArg({ type: "boolean" }, 1), /Boolean must be/);
  });
});

describe("validateArg - enum", () => {
  const def = { type: "enum", allowed: ["ping", "service", "syn"] };

  it("accepts allowed values", () => {
    assert.equal(validateArg(def, "ping"), "ping");
    assert.equal(validateArg(def, "syn"), "syn");
  });

  it("rejects disallowed values", () => {
    assert.throws(() => validateArg(def, "aggressive"), /not in allowed/);
  });
});

describe("validateArg - scope_target", () => {
  it("accepts valid targets", () => {
    assert.equal(validateArg({ type: "scope_target" }, "10.0.1.1"), "10.0.1.1");
    assert.equal(validateArg({ type: "scope_target" }, "example.com"), "example.com");
    assert.equal(
      validateArg({ type: "scope_target" }, "10.0.1.0/24"),
      "10.0.1.0/24"
    );
  });

  it("rejects wildcards", () => {
    assert.throws(() => validateArg({ type: "scope_target" }, "*.example.com"), /Wildcard/);
  });

  it("rejects injection", () => {
    assert.throws(() => validateArg({ type: "scope_target" }, "10.0.1.1;whoami"), /Injection/);
  });

  it("rejects leading or trailing whitespace", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, "example.com "),
      /whitespace/
    );
    assert.throws(
      () => validateArg({ type: "scope_target" }, " example.com"),
      /whitespace/
    );
    assert.throws(
      () => validateArg({ type: "scope_target" }, "\texample.com"),
      /whitespace/
    );
  });

  it("rejects values exceeding the 253-char RFC 1035 limit", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, "a".repeat(254)),
      /253/
    );
    // 4096-char buffer-pathological payload from cross-impl harness.
    assert.throws(
      () => validateArg({ type: "scope_target" }, "a".repeat(4096)),
      /253|exceeds/
    );
  });

  it("rejects empty input", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, ""),
      /empty/
    );
  });
});

describe("validateArg - url", () => {
  it("accepts valid URLs", () => {
    assert.equal(
      validateArg({ type: "url" }, "https://example.com"),
      "https://example.com"
    );
  });

  it("enforces scheme restrictions", () => {
    const def = { type: "url", schemes: ["https"] };
    assert.equal(validateArg(def, "https://example.com"), "https://example.com");
    assert.throws(() => validateArg(def, "http://example.com"), /scheme.*not in allowed/);
  });

  it("rejects invalid URLs", () => {
    assert.throws(() => validateArg({ type: "url" }, "not-a-url"), /Invalid URL/);
  });
});

describe("validateArg - path", () => {
  it("accepts valid relative paths", () => {
    assert.equal(validateArg({ type: "path" }, "tmp/wordlist.txt"), "tmp/wordlist.txt");
  });

  it("rejects absolute paths", () => {
    assert.throws(() => validateArg({ type: "path" }, "/tmp/wordlist.txt"), /must be relative/);
  });

  it("rejects path traversal", () => {
    assert.throws(() => validateArg({ type: "path" }, "etc/../shadow"), /traversal/);
  });

  it("rejects injection", () => {
    assert.throws(() => validateArg({ type: "path" }, "tmp/$(whoami)"), /Injection/);
  });
});

describe("validateArg - ip_address", () => {
  it("accepts valid IPv4", () => {
    assert.equal(validateArg({ type: "ip_address" }, "192.168.1.1"), "192.168.1.1");
  });

  it("accepts valid IPv6", () => {
    assert.equal(validateArg({ type: "ip_address" }, "::1"), "::1");
    assert.equal(validateArg({ type: "ip_address" }, "fe80::1"), "fe80::1");
  });

  it("rejects invalid IPs", () => {
    assert.throws(() => validateArg({ type: "ip_address" }, "999.999.999.999"), /Invalid IPv4/);
    assert.throws(() => validateArg({ type: "ip_address" }, "not-an-ip"), /Invalid IP/);
  });
});

describe("validateArg - cidr", () => {
  it("accepts valid CIDR", () => {
    assert.equal(validateArg({ type: "cidr" }, "10.0.0.0/24"), "10.0.0.0/24");
    assert.equal(validateArg({ type: "cidr" }, "192.168.1.0/16"), "192.168.1.0/16");
  });

  it("rejects invalid CIDR", () => {
    assert.throws(() => validateArg({ type: "cidr" }, "10.0.0.0"), /Invalid CIDR/);
    assert.throws(() => validateArg({ type: "cidr" }, "10.0.0.0/33"), /Invalid CIDR prefix/);
  });
});

describe("validateArg - number (float)", () => {
  it("accepts valid floats", () => {
    assert.equal(validateArg({ type: "number" }, "0.5"), 0.5);
    assert.equal(validateArg({ type: "number" }, "-3.14"), -3.14);
  });

  it("rejects NaN and infinity", () => {
    assert.throws(() => validateArg({ type: "number" }, "NaN"), /finite/);
    assert.throws(() => validateArg({ type: "number" }, "Infinity"), /finite/);
  });

  it("respects min_float / max_float", () => {
    const def = { type: "number", min_float: 0, max_float: 1 };
    assert.equal(validateArg(def, "0.5"), 0.5);
    assert.throws(() => validateArg(def, "-0.1"), /below minimum/);
    assert.throws(() => validateArg(def, "1.5"), /exceeds maximum/);
  });

  it("falls back to int min/max bounds", () => {
    const def = { type: "number", min: 1, max: 10 };
    assert.equal(validateArg(def, "5.5"), 5.5);
    assert.throws(() => validateArg(def, "0.5"), /below minimum/);
  });

  it("clamps when clamp=true", () => {
    const def = { type: "number", min_float: 0, max_float: 1, clamp: true };
    assert.equal(validateArg(def, "-5"), 0);
    assert.equal(validateArg(def, "5"), 1);
  });
});

describe("validateArg - scope_target hardening", () => {
  it("rejects punycode (xn--) labels", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, "xn--example-9c.com"),
      /punycode/
    );
    assert.throws(
      () => validateArg({ type: "scope_target" }, "sub.XN--example-9c.com"),
      /punycode/
    );
  });

  it("rejects non-ASCII (homoglyph IDN)", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, "exаmple.com"),
      /ASCII/
    );
  });

  it("surfaces specific traversal failure message", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, "../../etc/passwd"),
      /traversal/
    );
  });

  it("surfaces specific slash failure message", () => {
    assert.throws(
      () => validateArg({ type: "scope_target" }, "etc/passwd"),
      /'\/'/
    );
  });
});
